internal/common/context.bzl

```diff
@@ -0,0 +1,309 @@
+    "module_name": attr.string(),
+    "module_root": attr.string(),
+    # The official bazel rules look for this tag
+    "tags": ["NODE_MODULE_MARKER"],
+}
+
+###############################################################################
+# Helpers
+
+def _js_library_info(js, attr = None):
+    """Create a JsLibrary provider with attr.srcs and the sources from attr.deps
+
+    Args:
+    js: JsContext object
+    attr: rule attributes to extract srcs and deps from
+    """
+
+    if not attr:
+        attr = js.attr
+
+    # The srcs should contain what has been explicitly added for a rule
+    src_files = js._ctx.files.srcs
+
+    # The deps is list of labels that should have providers that we can get sources from
+    deps_attr = getattr(attr, "deps", [])
+
+    transitive_sources = []
+    node_modules_dirs = []
+    dep_modules = []
+    transitive_dep_modules = []
+    dep_module_targets = []
+    transitive_dep_module_targets = []
+
+    # Iterate through the deps to add them to their correct JsLibrary attributes
+    for dep in deps_attr:
+        if JsLibraryInfo in dep:
+            dep_js_library = dep[JsLibraryInfo]
+
+            # The dependency is another JsLibrary
+            transitive_sources.append(dep_js_library.all_sources)
+            transitive_dep_modules.append(dep_js_library.all_dep_modules)
+        elif JsModuleInfo in dep:
+            dep_js_module = dep[JsModuleInfo]
+            transitive_dep_modules.append(dep_js_module.all_dep_modules)
+            transitive_dep_module_targets.append(dep_js_module.all_dep_module_targets)
+        elif hasattr(dep, "tags") and "NODE_MODULES_MARKER" in getattr(dep, "tags"):
+            # The dependency is a module defined by rules_nodejs
+            direct_modules += dep
+
+        if JsModuleInfo in dep:
+            dep_js_module = dep[JsModuleInfo]
+            dep_module_targets.append(dep_js_module.module_root)
+            dep_modules.append(dep_js_module)
+
+        if NpmPackagesInfo in dep:
+            # The dependency is a node_modules directory installed by npm_packages
+            node_modules_dirs.append(dep[NpmPackagesInfo].installed_dir)
+
+    all_src_files = depset(
+        direct = src_files,
+        transitive = transitive_sources,
+    )
+
+    all_dep_modules = depset(
+        direct = dep_modules,
+        transitive = transitive_dep_modules,
+    )
+
+    all_dep_module_targets = depset(
+        direct = dep_module_targets,
+        transitive = transitive_dep_module_targets,
+    )
+
+    return JsLibraryInfo(
+        js = js,
+        package_path = js.package_path,
+        srcs = src_files,
+        all_sources = all_src_files,
+        node_modules_dirs = node_modules_dirs,
+        dep_modules = dep_modules,
+        all_dep_modules = all_dep_modules,
+        dep_module_targets = dep_module_targets,
+        all_dep_module_targets = all_dep_module_targets,
+    )
+
+def _library_to_source_info(js, library, gen_scripts = None):
```

Comment:
Would you mind adding some comments about `gen_scripts` here?

Reply:
Part of this is that I am still a little unsure about what the gen_scripts signature should look like. Currently the example is internal/ts_library/gsconfig.gen.js. The file exports a function as `module.exports` eg(with typescript types for some added clarity):
```ts
module.exports = async (ctx: {package: string, into: string, inputs: string[] }): Promise<{path: string, body: string}[]> => {}
```

Reply:
OK—that looks like a good start, we can always adjust as we go later.